File: dataset/parse_pgn.jl


using Redis
using ArgParse
using Distributed

using Chess
using Chess.PGN
using DataFrames
using JLD2

"""
    pop_chunk_from_list(conn::RedisConnection, list::Symbol,
    chunk_size::Int)::Vector{String}

Pops a chunk of data of size `chunk_size` from a list named `list` stored on
our redis server we are connected to with `conn`.
"""
function pop_chunk_from_list(
    conn::RedisConnection,
    list::Symbol,
    chunk_size::Int,
)::Vector{String}

    pipeline = open_pipeline(conn)
    # Gets values a chunk_sized subset of our list
    lrange(pipeline, list, 0, (chunk_size - 1))
    # Removes that same portion
    ltrim(pipeline, list, chunk_size, -1)

    data = read_pipeline(pipeline)[1]
    return data
end

function load_games(path::String)::Vector{String}

    conn = RedisConnection()

    file = open(path, "r")
    games::Vector{String} = []
    buffer::String = ""

    line = readline(file; keep=true)
    while line != nothing
        if line[1] == '['
            buffer = buffer * line
        end

        if line[1] == '1'
            buffer = buffer * line
            push!(games, buffer)
            buffer = ""
        end

        if length(line) > 2
            if (line[2] == '1') | (line[2] == '0')
                buffer = buffer * line
                push!(games, buffer)
                buffer = ""
            end
        end

        if length(games) == 1e6
            lpush(conn, :games, games)
            games = []
        end

        line = readline(file; keep=true)
    end
    lpush(conn, :games, games)
end

function parse_game(chunk::Int=100000)

    conn = RedisConnection()

    n = 0
    game_strs::Vector{String} = []
    while true

        game_strings = pop_chunk_from_list(conn, :games, chunk)

        if length(game_strings) == 0
            println(n)
            return
        end

        for game_string ∈ game_strings

            game = gamefromstring(game_string; annotations=true)

            if headervalue(game, "Termination") != "Normal"
                continue
            end

            try
                event = headervalue(game, "Event")
                white_elo = headervalue(game, "WhiteElo")
                black_elo = headervalue(game, "BlackElo")
                result = headervalue(game, "Result")
                move_str = string(game)[10:end]

                game_str = move_str * "," * result * "," * white_elo * "," * black_elo *","  * event
                push!(game_strs, game_str)

                if length(game_strs) == 1_000
                    lpush(conn, :data, game_strs)
                    game_strs = []
                end
            catch
                println("Missing data")
            end
        end
    end
    lpush(conn, :data, game_strs)
end

function to_dataset(file_path::String, chunk::Int=100000)

    conn = RedisConnection()

    file = open(file_path, "w")

    while true
        game_strings = pop_chunk_from_list(conn, :data, chunk)

        if length(game_strings) == 0
            return
        end

        for g in game_strings
            write(file, g * "\n")
        end
    end
end

function main()

    s = ArgParseSettings()
    @add_arg_table s begin
        "--load"
        help = ".pgn file to load into the job queue"
        "--parse"
        help = "Runs the worker process for the PGN -> FEN processing"
        action = :store_true
        "--to_dataset"
        help = "Converts the data into a dataset"
    end
    args = parse_args(s)

    file = args["load"]
    output_file = args["to_dataset"]

    if file != nothing
        load_games(file)
    elseif args["parse"]
        parse_game()
    elseif output_file != nothing
        to_dataset(output_file)
    end
end

main()

File: dataset/parse_pgn.py
import chess.pgn
import chess
import sys

def mirror_square(sq):
    row, col = sq[0], int(sq[1])
    return row + str(9 - col)

def mirror_move(m):
    move_str = str(m)
    from_sq, to_sq, promo = mirror_square(move_str[:2]), mirror_square(move_str[2:4]), move_str[4:]
    return chess.Move.from_uci(from_sq + to_sq + promo)


def parse_pgn(pgn, output_file):
    node = pgn.game()

    if 'Classical' not in pgn.headers['Event']:
        return

    result = pgn.headers['Result']
    if result not in ['0-1', '1-0']:
        return

    boards = []
    moves = []
    while node: 
        boards.append(node.board())
        node = node.next()
        if node:
            moves.append(node.move)
    
    boards = boards[:-1]
    if result == '1-0':
        boards = boards[0::2]
    else:
        boards = boards[1::2]

    for (b, m) in zip(boards, moves):

        if b.turn == chess.BLACK:
            b = b.mirror()
            m = mirror_move(m)
        
        output_file.write(f'{b.fen()},{str(m)}\n')

if __name__ == '__main__':
    path = sys.argv[1]
    pgn = open(path, "r")
    output_file = open(f'../data/{path[:-4]}.csv', "w")
    
    while True:
        pgn_file = chess.pgn.read_game(pgn)
        parse_pgn(pgn_file, output_file)
